datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -665,87 +656,81 @@ impl PhysicalExpr for BinaryExpr {
             children[1].clone(),
         )))
     }
-}
-
-impl PartialEq<dyn Any> for BinaryExpr {
-    fn eq(&self, other: &dyn Any) -> bool {
-        down_cast_any_ref(other)
-            .downcast_ref::<Self>()
-            .map(|x| self.left.eq(&x.left) && self.op == x.op && self.right.eq(&x.right))
-            .unwrap_or(false)
-    }
-}
-
-struct BinaryExprStats {
-    op: Operator,
-    left: Arc<dyn PhysicalExpr>,
-    right: Arc<dyn PhysicalExpr>,
-}
 
-impl PhysicalExprStats for BinaryExprStats {
-    fn boundaries(&self, columns: &[ColumnStatistics]) -> Option<ExprBoundaries> {
+    /// Return the boundaries of this binary expression's result. If the expression itself
+    /// is a comparison which changes the boundaries of one of its inputs (a = 20 would pin
+    /// a to 20), then it might update the input's boundaries directly on the context.
+    fn analyze(&self, context: &mut AnalysisContext) -> Option<ExprBoundaries> {
```

Comment:
Since AnalysisContext is in terms of the "current schema" it will likely almost always end up being modified if the execution plan modifies the schema (which most do).

I find that the mix of "input" and "output" of a `mut &AnalysisContext` to be confusing to explain.


Since an `AnalysisContext` has `ExprBoundaries` in it anyways, I recommend the signature be


```rust
fn analyze(&self) -> AnalysisContext {
...
}
```

And then pass each input can produce the context / boundary analysis for itself and the parent operators can combine them

Given that the inputs will produced an owned AnalysisContext it will be clear what is input and what is output and I think it can be made quite fast in Rust

Reply:
Maybe AnalysisContext could be expanded to be:

```rust
/// A context for collecting and aggregating known boundaries of an expression
/// tree.
#[derive(Clone, Debug, PartialEq)]
pub struct AnalysisContext {
    /// the boundaries of this particular PhysicalExpr's output
    pub boundaries: ExprBoundaries,

    /// A list of known column boundaries, ordered by column index
    /// in the current schema.
    pub column_boundaries: Vec<Option<ExprBoundaries>>,
}
```

As I have mentioned before, I don't fully understand specially how `column_boundaries` can be used, but I am interested in seeing an example

Reply:
> Since AnalysisContext is in terms of the "current schema" it will likely almost always end up being modified if the execution plan modifies the schema (which most do).

You can think about as if it were a Statistics for that execution plan. It will be generated by the same place and will be constructed from the existing statistics (at least in `FilterExec`).

> Since an AnalysisContext has ExprBoundaries in it anyways, I recommend the signature be
> fn analyze(&self) -> AnalysisContext {

If I understood it correctly, the problem with this approach is that it won't be able to see the existing boundaries of any columns. The place that wants to do the analysis needs to pass column information (otherwise it is not going to be a very useful analysis; unless the expression itself is self-sufficient `1 > 2`). All the children also need to receive a context, but depending on the parent (`OR` for example) it might be a different one than the one we received initially.

> As I have mentioned before, I don't fully understand specially how column_boundaries can be used, but I am interested in seeing an example

The only 'unknown' during the analysis process is the column references. Anything else can be inferred by the behaviour of the particular expression itself. So we need a way of conveying that information to each sub-expression when doing the analysis. I initially started using a vector of column statistics but then once we had to change it on certain expressions, like for example `a > 20 AND a < 70` where we have different boundaries for `a` when evaluating the second part of the conjunction (`a < 70`), I went with a mutable/shared context.

Reply:
> Since AnalysisContext is in terms of the "current schema" it will likely almost always end up being modified if the execution plan modifies the schema (which most do).
> 
> I find that the mix of "input" and "output" of a `mut &AnalysisContext` to be confusing to explain.
> 
> Since an `AnalysisContext` has `ExprBoundaries` in it anyways, I recommend the signature be
> 
> ```rust
> fn analyze(&self) -> AnalysisContext {
> ...
> }
> ```
> 
> And then pass each input can produce the context / boundary analysis for itself and the parent operators can combine them
> 
> Given that the inputs will produced an owned AnalysisContext it will be clear what is input and what is output and I think it can be made quite fast in Rust

I would prefer this change also. It makes the signature more clear. It is similar to the `EquivalenceProperties` in the ExecutionPlan, mixing "input" and "output" is very confusing.